### ISSUE TITLE
neur

### ISSUE THREAD

**Issue by Zvonok13** [Zvonok13/Trash_sort#1]
import matplotlib.pyplot as plt
import numpy as np
import PIL
import tensorflow as tf
import pathlib

from tensorflow.keras import layers

# Укажите путь к вашим данным
data_dir = pathlib.Path('C:/Users/zvono/PycharmProjects/od arduino/dataset')
image_count = len(list(data_dir.glob('*/*.jpg')))
print(image_count)

# Загрузка изображений
metal = list(data_dir.glob('metal/*'))
PIL.Image.open(str(metal[0]))
plastic = list(data_dir.glob('plastic/*'))
PIL.Image.open(str(plastic[0]))
PIL.Image.open(str(plastic[1]))
PIL.Image.open(str(metal[1]))

# Параметры для обучения
batch_size = 32
img_height = 180
img_width = 180

# Создание обучающего и проверочного набора данных
train_ds = tf.keras.utils.image_dataset_from_directory(
    data_dir,
    validation_split=0.2,
    subset="training",
    seed=123,
    image_size=(img_height, img_width),
    batch_size=batch_size)

val_ds = tf.keras.utils.image_dataset_from_directory(
    data_dir,
    validation_split=0.2,
    subset="validation",
    seed=123,
    image_size=(img_height, img_width),
    batch_size=batch_size)

class_names = train_ds.class_names
print(class_names)

# Нормализация данных
normalization_layer = layers.Rescaling(1. / 255)

# Создание модели
model = tf.keras.Sequential([
    layers.experimental.preprocessing.Rescaling(1./255, input_shape=(img_height, img_width, 3)),
    layers.Conv2D(16, 3, padding='same', activation='relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(32, 3, padding='same', activation='relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(64, 3, padding='same', activation='relu'),
    layers.MaxPooling2D(),
    layers.Dropout(0.6),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(len(class_names), activation='softmax')  # Последний слой с Softmax
])

# Компиляция модели
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Обучение модели
epochs = 50
history = model.fit(
    train_ds,
    validation_data=val_ds,
    epochs=epochs
)

# Визуализация метрик обучения
acc = history.history['accuracy']
val_acc = history.history['val_accuracy']
loss = history.history['loss']
val_loss = history.history['val_loss']
epochs_range = range(epochs)

plt.figure(figsize=(8, 8))
plt.subplot(1, 2, 1)
plt.plot(epochs_range, acc, label='Training Accuracy')
plt.plot(epochs_range, val_acc, label='Validation Accuracy')
plt.legend(loc='lower right')
plt.title('Training and Validation Accuracy')

plt.subplot(1, 2, 2)
plt.plot(epochs_range, loss, label='Training Loss')
plt.plot(epochs_range, val_loss, label='Validation Loss')
plt.legend(loc='upper right')
plt.title('Training and Validation Loss')
plt.show()
# Сохранение модели
model.save('garbage_recognition_model_au7110.h5')
